### DISCUSSION TITLE
[ZK Stack] - Instructions to Deploy a ZkStack Chain with Custom Token as a Gas token

### DISCUSSION THREAD

**Discussion by brucewhine** [zkSync-Community-Hub/zksync-developers#816] [ZK Stack] [answered]
### Team or Project

Gotham

### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [ ] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [X] Other (please specify below)

### Rust Version

1.75.0

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

I'm trying to create a custom zk chain, but after creating the ecosystem and trying to initialise the chain, I'm getting the following error.

● Initializing chain xxx
│
■ Failed to create portal config
│
▲ 0: Invalid name: please ensure the contract and method you're calling exist! failed to decode empty bytes. if you're using jsonrpc this is likely due to jsonrpc returning 0x in case contract or method don't exist
│
└ Failed to run command

Followed so far, 
For setting up the custom token I've followed this doc: https://code.zksync.io/tutorials/custom-zk-chain/customizing-your-chain.

I've used the governer wallet to deploy the token contract on L1 and used the same address while creating the chain, but when doing
zkstack chain init
getting an insufficient balance error from a different address other than the governor account.

If I send the balance to that account and try again, getting BridgeHubAlreadyRegistered Error

### Expected Behavior

Custom token to be Gas token for the Rollup 

### Current Behavior

Getting error setting up the Custom token 

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
It seems like you are encountering multiple issues while setting up a custom zk chain with a custom token as the gas token. Here are some steps and considerations based on the information you've provided:

1. **Invalid Name Error**: This error typically indicates that the contract or method name provided does not exist or is not correctly deployed. Double-check the contract address and method names in your deployment scripts and ensure they are correctly deployed on L1.

2. **Insufficient Balance Error**: This error suggests that the address trying to initialize the chain does not have enough balance of the custom token to pay for the gas fees. Ensure that the address you are using to initialize the chain has a sufficient balance of the custom token. You might need to transfer an adequate amount of the custom token to this address.

3. **BridgeHubAlreadyRegistered Error**: This error occurs if you are trying to register a token or address that has already been registered in the BridgeHub contract. Ensure that the token or address you are trying to register is not already registered. If it is, you may need to use a different address or token, or deregister the existing entry if possible.

For detailed guidance on setting up a custom zk chain with a custom ERC20 token as the gas token, you can refer to the [Custom Base Tokens documentation](https://docs.zksync.io/zk-stack/running-a-zk-chain/custom-base-tokens). This guide provides comprehensive steps for deploying an ERC20 token to L1, creating a new chain that uses this ERC20 token, and setting it as the default.

If these steps do not resolve your issues, it may be helpful to revisit the configuration files and deployment scripts to ensure all parameters and addresses are correctly set. Additionally, consider reaching out on the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions) for more in-depth support.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
Did you make sure to initialize the ecosystem too? https://code.zksync.io/tutorials/custom-zk-chain#deploying-the-ecosystem

I went through the tutorial and was able to get everything running. Had to change a few of the ports but ultimately got 2 chains deployed and one with a custom base token.

**Comment by sachanayush47** [comment, marked as answer]
To fix `BridgeHubAlreadyRegistered` error, update the chain ID to something else. For example, if your chain ID is 456, change it to a different value. This error occurred because in your 1st attempt chain has been successfully registered to the Bridge Hub and you cannot register another chain with same ID.

**Comment by monk07-01** [comment]
Hi All I am trying to deploy a custom base token chain as well but getting errors. As per the tutorial and docs it only has option for local reth node but i choosed Sepolia for L1 chain and not able to build. So do we need to follow both the tutorials together as you cannot create a new chain without ecosystem but ecosystem creates a new chain itself but the tweaks and twists become difficult if using different environment like Sepolia. Please advise what is the best way to go that way. I usually get stuck after creating ecosystem command. I am able to do it properly. It is confusing for me of how to bridge funds and also how to add the token number . error {{
$ zkstack ecosystem init --dev

┌   ZK Stack CLI 
│
●  Initializing ecosystem
│  
◇  Downloading era observability... done in 0.401952138 secs
│
◇  Installing and building dependencies... done in 546.170377462 secs
│
◐  Deploying ecosystem contracts...                                                                                                                                                                                                              Command: forge script deploy-scripts/DeployL1.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --slow --private-key=7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110 --broadcast
◇  Deploying ecosystem contracts... done in 19.714809931 secs
│
◒  Accepting governance...                                                                                                                                                                                                                       Command: forge script deploy-scripts/AcceptAdmin.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --sig=83eddd19000000000000000000000000fbbe8328e6d5e7c899bdfee9386639f38ee0ced700000000000000000000000023a7bf17d6883eb29771d59afb64a0f3e437585c --private-key=
◇  Accepting governance... done in 3.281986283 secs
│
◒  Accepting governance...                                                                                                                                                                                                                       Command: forge script deploy-scripts/AcceptAdmin.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --sig=7475e9ea0000000000000000000000005994b24b317d3401cd20ed78844eaaea6e5b003d00000000000000000000000023a7bf17d6883eb29771d59afb64a0f3e437585c --private-key=
◇  Accepting governance... done in 1.010525872 secs
│
◒  Accepting governance...                                                                                                                                                                                                                       Command: forge script deploy-scripts/AcceptAdmin.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --sig=83eddd19000000000000000000000000fbbe8328e6d5e7c899bdfee9386639f38ee0ced7000000000000000000000000a789d6c2c96fc8a498448cad862965b31cc92115 --private-key=
◇  Accepting governance... done in 1.017359985 secs
│
◒  Accepting governance...                                                                                                                                                                                                                       Command: forge script deploy-scripts/AcceptAdmin.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --sig=83eddd19000000000000000000000000fbbe8328e6d5e7c899bdfee9386639f38ee0ced70000000000000000000000009be8a2d20f3070157741ef8bda5fed8ba037c34b --private-key=
◇  Accepting governance... done in 1.017034788 secs
│
◒  Accepting governance...                                                                                                                                                                                                                       Command: forge script deploy-scripts/AcceptAdmin.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --sig=7475e9ea0000000000000000000000005994b24b317d3401cd20ed78844eaaea6e5b003d0000000000000000000000009be8a2d20f3070157741ef8bda5fed8ba037c34b --private-key=
◇  Accepting governance... done in 1.012617523 secs
│
◒  Accepting governance...                                                                                                                                                                                                                       Command: forge script deploy-scripts/AcceptAdmin.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --sig=83eddd19000000000000000000000000fbbe8328e6d5e7c899bdfee9386639f38ee0ced70000000000000000000000005a966733e6fc80ae96a38b7f8e1527ef6a3e242c --private-key=
◇  Accepting governance... done in 1.02096671 secs
│
●  Deploying ERC20 contracts
│  
◐  Deploying ERC20 contracts...                                                                                                                                                                                                                  Command: forge script deploy-scripts/DeployErc20.s.sol --legacy --ffi --rpc-url=http://127.0.0.1:8545 --broadcast --private-key=
◇  Deploying ERC20 contracts... done in 1.015580691 secs
│
●  Initializing chain bfh_testnet
│  
■  Failed to create portal config
│  
▲    0: Invalid name: please ensure the contract and method you're calling exist! failed to decode empty bytes. if you're using jsonrpc this is likely due to jsonrpc returning `0x` in case contract or method don't exist
│  
└  Failed to run command}

**Reply by sachanayush47** [reply]
Share your social media handle. Will reach out to you and help you to deploy.

**Reply by monk07-01** [reply]
Hi @sachanayush47 , I am able to deploy my custom token chain using rethnode but I fail to bridge as I am confused what private key to use. In the custom token section in docs it shows keys of L1 account with rich eth which seems incorrect as my chain is custom token. If I try to send via contract I need a private key to sign but cant use the governor added from L1 wallet. Kindly advise what to do?

**Reply by sachanayush47** [reply]
Use token owner private key.

**Reply by monk07-01** [reply]
but that is the main question, before creating a custom token chain, I created a zk_test_1 and the ecosystem as per docs.https://code.zksync.io/tutorials/custom-zk-chain. After that before creating the second custom token chain we need to deploy the contract in L1 for which I used the same key as governor of L1 chain created. When I use this same wallet private key, it gives me an error when bridging my tokens on L2. Also please let me know how to do it using sepolia.
